linera-sdk/src/lib.rs

```diff
@@ -366,7 +366,32 @@ pub struct QueryContext {
     pub chain_id: ChainId,
 }
 
-/// A message together with routing information.
+/// Resources that an application may spend during the execution of a message.
+#[derive(Clone, Copy, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+pub struct Resources {
+    /// An amount of execution fuel.
+    pub fuel: u64,
+    /// A number of read operations to be executed.
+    pub read_operations: u32,
+    /// A number of write operations to be executed.
+    // TODO(#1530): This is not used at the moment.
+    pub write_operations: u32,
+    /// A number of bytes to read.
+    pub bytes_to_read: u32,
+    /// A number of bytes to write.
+    pub bytes_to_write: u32,
+    /// A number of messages to be sent.
+    pub messages: u32,
+    /// The size of the messages to be sent.
+    // TODO(#1531): Account for the type of message to be sent.
+    pub message_size: u32,
+    /// An increase in the amount of storage space.
+    pub storage_size_delta: u32,
+    // TODO(#1532): Account for the system calls that we plan on calling.
+    // TODO(#1533): Allow declaring calls to other applications instead of having to count them here.
```

Comment:
We could just implement `Add` for `Resources`, and then use addition to merge resource budgets :thinking:

Reply:
The current proposal in #1533 uses less fuel, I think, but we can discuss there.